receiver/snmpreceiver/client.go

```diff
@@ -0,0 +1,539 @@
+	case "AUTH_PRIV":
+		(*client).SetMsgFlags(gosnmp.AuthPriv)
+		authProtocol, err := getAuthProtocol(cfg.AuthType)
+		if err != nil {
+			return err
+		}
+		securityParams.AuthenticationProtocol = authProtocol
+		securityParams.AuthenticationPassphrase = cfg.AuthPassword
+
+		privProtocol, err := getPrivacyProtocol(cfg.PrivacyType)
+		if err != nil {
+			return err
+		}
+		securityParams.PrivacyProtocol = privProtocol
+		securityParams.PrivacyPassphrase = cfg.PrivacyPassword
+	default:
+		return fmt.Errorf("invalid security protocol '%s'", cfg.SecurityLevel)
+	}
+	(*client).SetSecurityParameters(securityParams)
+
+	return nil
+}
+
+// getAuthProtocol gets gosnmp auth protocol based on config auth type
+func getAuthProtocol(authType string) (gosnmp.SnmpV3AuthProtocol, error) {
+	switch strings.ToUpper(authType) {
+	case "MD5":
+		return gosnmp.MD5, nil
+	case "SHA":
+		return gosnmp.SHA, nil
+	case "SHA224":
+		return gosnmp.SHA224, nil
+	case "SHA256":
+		return gosnmp.SHA256, nil
+	case "SHA384":
+		return gosnmp.SHA384, nil
+	case "SHA512":
+		return gosnmp.SHA512, nil
+	default:
+		return gosnmp.MD5, fmt.Errorf("invalid auth protocol '%s'", authType)
+	}
+}
+
+// getPrivacyProtocol gets gosnmp privacy protocol based on config privacy type
+func getPrivacyProtocol(privacyType string) (gosnmp.SnmpV3PrivProtocol, error) {
+	switch strings.ToUpper(privacyType) {
+	case "DES":
+		return gosnmp.DES, nil
+	case "AES":
+		return gosnmp.AES, nil
+	case "AES192":
+		return gosnmp.AES192, nil
+	case "AES192C":
+		return gosnmp.AES192C, nil
+	case "AES256":
+		return gosnmp.AES256, nil
+	case "AES256C":
+		return gosnmp.AES256C, nil
+	default:
+		return gosnmp.DES, fmt.Errorf("invalid privacy protocol '%s'", privacyType)
+	}
+}
+
+// Connect uses the goSNMP client's connect
+func (c *snmpClient) Connect() error {
+	return c.client.Connect()
+}
+
+// Close uses the goSNMP client's close
+func (c *snmpClient) Close() error {
+	return c.client.Close()
+}
+
+// GetScalarData retrieves scalar metrics from passed in scalar OIDs. The returned data
+// is then also passed into the provided function.
+// Note: These OIDs must all end in ".0" for the SNMP GET to work correctly
+func (c *snmpClient) GetScalarData(oids []string, processFn processFunc) error {
+	// Nothing to do if there are no OIDs
+	if len(oids) == 0 {
+		return nil
+	}
+
+	// Group OIDs into chunks based on the max amount allowed in a single SNMP GET
+	chunkedOIDs := chunkArray(oids, c.client.GetMaxOids())
+	getOIDsSuccess := false
+	processOIDSuccess := false
+
+	// For each group of OIDs
+	for _, oids := range chunkedOIDs {
+		// Note: Not implementing GetBulk as I don't think it would work correctly for the current design
+		packets, err := c.client.Get(oids)
+		if err != nil {
+			c.logger.Warn("Problem with GET OIDs", zap.Error(err))
+			// Prevent getting stuck in a failure where we can't recover
+			if strings.Contains(err.Error(), "request timeout (after ") {
```

Comment:
Does SNMP have specific errors that we could do an `errors.Is` on rather than a string comparison?

Reply:
@cpheps i think this is the only specific gosnmp error that i look at. i'll look into this

Reply:
@cpheps  looks like there's not a specific error i can check against.
`err = fmt.Errorf("request timeout (after %d retries)", retries-1)`